File: dist-scripts/test/utils/CallScenarioDetector.js
"use strict";
/**
 * CallScenarioDetector.ts
 * TypeScript port of Java CallScenarioDetector — automated call outcome classification
 * for reporting and user-facing comments in calling tests.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectCallScenario = detectCallScenario;
exports.getScenarioEmoji = getScenarioEmoji;
function containsIgnoreCase(source, ...keywords) {
    if (!source)
        return false;
    const lower = source.toLowerCase();
    return keywords.some(kw => lower.includes(kw.toLowerCase()));
}
/**
 * Detect call scenario from UiAutomator page source and call metrics.
 *  CallScenarioDetector.detectCallScenario().
 */
function detectCallScenario(pageSource, callConnected, actualDuration, targetDuration) {
    if (containsIgnoreCase(pageSource, 'ringing', 'ring') && callConnected) {
        return {
            scenarioNumber: 1,
            scenarioName: 'Call Ringing',
            status: 'CONNECTED',
            comment: `Call: Connected (${actualDuration}s) | Ringing detected`
        };
    }
    if (containsIgnoreCase(pageSource, 'busy')) {
        return {
            scenarioNumber: 2,
            scenarioName: 'Busy Tone',
            status: 'CONNECTED (BUSY)',
            comment: `Call: Busy tone detected (${actualDuration}s)`
        };
    }
    if (callConnected && actualDuration >= targetDuration * 0.9) {
        return {
            scenarioNumber: 3,
            scenarioName: 'Call Connected',
            status: 'CONNECTED',
            comment: `Call: Connected (${actualDuration}s)`
        };
    }
    if (callConnected &&
        actualDuration < targetDuration &&
        containsIgnoreCase(pageSource, 'drop', 'disconnect')) {
        return {
            scenarioNumber: 4,
            scenarioName: 'Call Dropped',
            status: 'CONNECTED (DROPPED)',
            comment: `Call dropped after ${actualDuration}s`
        };
    }
    if (containsIgnoreCase(pageSource, 'invalid number', 'invalid_number', 'format incorrect')) {
        return {
            scenarioNumber: 5,
            scenarioName: 'Invalid Number',
            status: 'FAILED',
            comment: 'Invalid number format detected'
        };
    }
    if (containsIgnoreCase(pageSource, 'network') &&
        containsIgnoreCase(pageSource, 'error', 'busy', 'unavailable')) {
        return {
            scenarioNumber: 6,
            scenarioName: 'Network Error',
            status: 'FAILED',
            comment: 'No network signal detected'
        };
    }
    if (containsIgnoreCase(pageSource, 'not reachable', 'out of coverage', 'unreachable', 'switched off')) {
        return {
            scenarioNumber: 7,
            scenarioName: 'Not Reachable',
            status: 'FAILED',
            comment: 'Call failed - B party not reachable'
        };
    }
    if (containsIgnoreCase(pageSource, 'outgoing') &&
        containsIgnoreCase(pageSource, 'barred', 'restricted')) {
        return {
            scenarioNumber: 8,
            scenarioName: 'A-Party Barred',
            status: 'FAILED',
            comment: 'Call failed - Outgoing barred'
        };
    }
    if (containsIgnoreCase(pageSource, 'incoming') &&
        containsIgnoreCase(pageSource, 'barred', 'restricted')) {
        return {
            scenarioNumber: 9,
            scenarioName: 'B-Party Barred',
            status: 'REJECTED',
            comment: 'Call rejected - Incoming barred'
        };
    }
    if (containsIgnoreCase(pageSource, 'wrong sim', 'wrong_sim', 'sim error', 'no sim')) {
        return {
            scenarioNumber: 10,
            scenarioName: 'Wrong SIM',
            status: 'FAILED',
            comment: 'Call failed due to wrong SIM slot'
        };
    }
    if (callConnected && actualDuration >= targetDuration * 0.8) {
        return {
            scenarioNumber: 11,
            scenarioName: 'Call Quality',
            status: 'CONNECTED',
            comment: `Call: Connected (${actualDuration}s) | Audio quality OK`
        };
    }
    if (containsIgnoreCase(pageSource, 'crbt', 'caller tune', 'ring back tone') &&
        callConnected) {
        return {
            scenarioNumber: 12,
            scenarioName: 'CRBT Detected',
            status: 'CONNECTED',
            comment: `Call: Connected with CRBT detected (${actualDuration}s)`
        };
    }
    return {
        scenarioNumber: 0,
        scenarioName: 'Unknown',
        status: 'FAILED',
        comment: 'Call scenario could not be determined'
    };
}
function getScenarioEmoji(scenarioNumber) {
    var _a;
    const map = {
        1: '🔔',
        2: '📞',
        3: '',
        4: '',
        5: '',
        6: '📡',
        7: '📵',
        8: '🚫',
        9: '🚫',
        10: '',
        11: '📊',
        12: '🎵'
    };
    return (_a = map[scenarioNumber]) !== null && _a !== void 0 ? _a : '❓';
}

File: src/main/java/com/telecom/utils/CallScenarioDetector.java
package com.telecom.utils;

/**
 * Automated detection of call scenarios
 */
public class CallScenarioDetector {
    
    public static class CallScenario {
        private String scenarioName;
        private String status;
        private String comment;
        private int scenarioNumber;
        
        public CallScenario(int number, String name, String status, String comment) {
            this.scenarioNumber = number;
            this.scenarioName = name;
            this.status = status;
            this.comment = comment;
        }
        
        // Getters
        public String getScenarioName() { return scenarioName; }
        public String getStatus() { return status; }
        public String getComment() { return comment; }
        public int getScenarioNumber() { return scenarioNumber; }
    }
    
    /**
     * Main method to detect call scenario from page source
     */
    public static CallScenario detectCallScenario(String pageSource, boolean callConnected, 
                                                   int actualDuration, int targetDuration) {
        
        // Scenario 1: Call Ringing
        if (containsIgnoreCase(pageSource, "ringing", "ring") && callConnected) {
            return new CallScenario(1, "Call Ringing", "CONNECTED", 
                String.format("Call: Connected (%ds) | Ringing detected", actualDuration));
        }
        
        // Scenario 2: Busy Tone
        if (containsIgnoreCase(pageSource, "busy")) {
            return new CallScenario(2, "Busy Tone", "CONNECTED (BUSY)", 
                String.format("Call: Busy tone detected (%ds)", actualDuration));
        }
        
        // Scenario 3: Call Connected (default successful scenario)
        if (callConnected && actualDuration >= targetDuration * 0.9) {
            return new CallScenario(3, "Call Connected", "CONNECTED", 
                String.format("Call: Connected (%ds)", actualDuration));
        }
        
        // Scenario 4: Call Dropped
        if (callConnected && actualDuration < targetDuration && 
            containsIgnoreCase(pageSource, "drop", "disconnect")) {
            return new CallScenario(4, "Call Dropped", "CONNECTED (DROPPED)", 
                String.format("Call dropped after %ds", actualDuration));
        }
        
        // Scenario 5: Invalid Number
        if (containsIgnoreCase(pageSource, "invalid number", "invalid_number", 
                              "format incorrect")) {
            return new CallScenario(5, "Invalid Number", "FAILED", 
                "Invalid number format detected");
        }
        
        // Scenario 6: Network Error
        if (containsIgnoreCase(pageSource, "network") && 
            containsIgnoreCase(pageSource, "error", "busy", "unavailable")) {
            return new CallScenario(6, "Network Error", "FAILED", 
                "No network signal detected");
        }
        
        // Scenario 7: Not Reachable
        if (containsIgnoreCase(pageSource, "not reachable", "out of coverage", 
                              "unreachable", "switched off")) {
            return new CallScenario(7, "Not Reachable", "FAILED", 
                "Call failed - B party not reachable");
        }
        
        // Scenario 8: A-Party Barred
        if (containsIgnoreCase(pageSource, "outgoing") && 
            containsIgnoreCase(pageSource, "barred", "restricted")) {
            return new CallScenario(8, "A-Party Barred", "FAILED", 
                "Call failed - Outgoing barred");
        }
        
        // Scenario 9: B-Party Barred
        if (containsIgnoreCase(pageSource, "incoming") && 
            containsIgnoreCase(pageSource, "barred", "restricted")) {
            return new CallScenario(9, "B-Party Barred", "REJECTED", 
                "Call rejected - Incoming barred");
        }
        
        // Scenario 10: Wrong SIM
        if (containsIgnoreCase(pageSource, "wrong sim", "wrong_sim", 
                              "sim error", "no sim")) {
            return new CallScenario(10, "Wrong SIM", "FAILED", 
                "Call failed due to wrong SIM slot");
        }
        
        // Scenario 11: Call Quality (connected with good quality)
        if (callConnected && actualDuration >= targetDuration * 0.8) {
            return new CallScenario(11, "Call Quality", "CONNECTED", 
                String.format("Call: Connected (%ds) | Audio quality OK", actualDuration));
        }
        
        // Scenario 12: CRBT Detected
        if (containsIgnoreCase(pageSource, "crbt", "caller tune", "ring back tone") && 
            callConnected) {
            return new CallScenario(12, "CRBT Detected", "CONNECTED", 
                String.format("Call: Connected with CRBT detected (%ds)", actualDuration));
        }
        
        // Default: Generic failure
        return new CallScenario(0, "Unknown", "FAILED", 
            "Call scenario could not be determined");
    } 
    
    /**
     * Helper method for case-insensitive multi-keyword search
     */
    private static boolean containsIgnoreCase(String source, String... keywords) {
        if (source == null) return false;
        String lowerSource = source.toLowerCase();
        for (String keyword : keywords) {
            if (lowerSource.contains(keyword.toLowerCase())) {
                return true;
            }
        }
        return false;
    }
    
    /**
     * Get scenario emoji for reporting
     */
    public static String getScenarioEmoji(int scenarioNumber) {
        switch (scenarioNumber) {
            case 1: return "✅"; // Ringing
            case 2: return "📞"; // Busy
            case 3: return "✅"; // Connected
            case 4: return "⚠️"; // Dropped
            case 5: return "❌"; // Invalid
            case 6: return "📡"; // Network Error
            case 7: return "📵"; // Not Reachable
            case 8: return "🚫"; // A-Party Barred
            case 9: return "🚫"; // B-Party Barred
            case 10: return "📱"; // Wrong SIM
            case 11: return "📊"; // Call Quality
            case 12: return "🎵"; // CRBT
            default: return "❓";
        }
    }
}
